### ISSUE TITLE
`;` is unexpected if it after function body

### ISSUE THREAD

**Issue by fzzr-** [TiVo/intellij-haxe#110]
See screenshot for example:

![screen shot 2015-01-20 at 13 25 58](https://cloud.githubusercontent.com/assets/888526/5817509/98171042-a0bd-11e4-83b8-855e6bbe6cbd.png)

Example code for reproduce:

````haxe
class Moo
{
	// `;` - it's ok for compiler but not for Idea:
	public function new(){};
	function func():Void {};

	// here all ok:
	function foo():Void {}
	function obj() return {};
	function fool():Void null;
	function bar():Void return;
}
````

**Comment by EBatTiVo** [comment]
Using IDEA 13.1.5 (135.1286) and plugin build 0.8.1.1.TiVo.ClassHierarchy.11, I do not see the issue.  The example code shows no errors in IDEA and it compiles just fine.

With what configuration and build are you seeing this?

**Comment by fzzr-** [comment]
Latest Idea 14 EAP.

**Comment by as3boyan** [comment]
I think in plugin `;` allowed only for extern members currently

**Comment by Simn** [comment]
@as3boyan: Why does `extern` influence parsing at all? Our own parser does not care about it because all extern-related errors are typing errors. Extern classes should be parsed exactly as non-extern ones.

**Comment by as3boyan** [comment]
@EBatTiVo I was able to reproduce this issue on latest

@Simn I think that is how it was implemented by Fedor I think

**Comment by as3boyan** [comment]
It seems like according to Haxe compiler(try.haxe.org)
For `extern` classes, function body is optional.
For regular classes, function body is required

**Comment by EBatTiVo** [comment]
I think Srikanth may have fixed this last week.  He was working on some other stuff and found errors with how external functions were parsing, so he tweaked those rules.

**Comment by Simn** [comment]
@as3boyan: Yes but these are not parser differences. They are typing errors which you shouldn't have to worry about.

**Comment by EBatTiVo** [comment]
Showing typing errors are what IDEs are all about.  *N'est ces pas?*

**Comment by EBatTiVo** [comment]
Also, Srikanth was NOT working in IDES 14, so the bug may not have been fixed there.

**Comment by Simn** [comment]
You can show errors at your leisure but this should _not_ affect parsing in any way. Getting something like "Unexpected ;" for this isn't exactly quality error reporting anyway so the IDE should just tell the user what the actual problem is.

**Comment by as3boyan** [comment]
@Simn You right, it should be parsed anyway. With or without body. Like we used `(functionCommonBody | ';')` for regular functions

**Comment by as3boyan** [comment]
@simn Agreed, it shouldn't break parsing.

**Comment by as3boyan** [comment]
Or maybe even `functionCommonBody? ';'?` need to look into Haxe compiler how it's done there.

**Comment by EBatTiVo** [comment]
So, I think we may be looking at this a bit wrong.  The parser isn't breaking; the rest of the file continues to be parsed.  

What we're getting here is simply that the parser found the semi-colon /after/ the function rule was completed.  (And I would argue that the message "unexpected ';'," while not necessarily useful, is exactly what the problem is here.)  If the haxe compiler does not (and will not) complain about a spurious semi-colon appearing in class definitions (because the semi-colon is not part of the function definition), then we should loosen up the BNF to allow them anywhere in a class definition.  (e.g. add an empty statement rule).

I'm not an OCAML jock (yet), so I haven't clawed my way through the compiler's parser to see how it handles this.  @Simn, is the semi-colon as given in the example above, considered part of the function rule, or a separate statement?

**Comment by Simn** [comment]
@EBatTiVo: After the function header (the optional type-hint) the parser expects either an expression or a semicolon. If there's an expression it also allows an additional optional comma. So the rule is something like `(';' | FunctionBody ';'?)`

**Comment by as3boyan** [comment]
Good point, then we could just allow additional semi-colon:

```
externFunctionDeclaration ::= (functionMacroMember| declarationAttribute)* 'function' componentName genericParam? '(' parameterList ')' typeTag? 'untyped'? (functionCommonBody ';'? | ';')
{pin=3 mixin="com.intellij.plugins.haxe.lang.psi.impl.HaxeMethodImpl" implements="com.intellij.plugins.haxe.lang.psi.HaxeMethod"}
functionDeclarationWithAttributes ::= (functionMacroMember | declarationAttribute)* 'function' (constructorName | componentName) genericParam? '(' parameterList ')' typeTag? 'untyped'? functionCommonBody ';'?
{pin=3 mixin="com.intellij.plugins.haxe.lang.psi.impl.HaxeMethodImpl" implements="com.intellij.plugins.haxe.lang.psi.HaxeMethod"}
```

**Comment by EBatTiVo** [comment]
Well, not quite that easy.  But pretty close.  The rules some careful consideration because some of them already allow the semi-colon, and just adding one, or (as in the case of externFunctionDeclaration, making the functionCommonBody optional) may make the rules ambiguous (more than one rule matches a single function definition).

```
externFunctionDeclaration ::= (functionMacroMember| declarationAttribute)* 'function' componentName genericParam? '(' parameterList ')' typeTag? 'untyped'? (functionCommonBody | ';')
```

**Comment by Simn** [comment]
The `untyped` is out of place there. It should be part of the expression parsing.

But I wasn't entirely precise either. The optional semicolon is only allowed for block expressions, i.e. if the expression ended with `}`.

**Comment by as3boyan** [comment]
So then it's a completely different story, semi-colon then might be allowed on block expressions... (blockStatement)...

**Comment by EBatTiVo** [comment]
I don't think that it's all block expressions, just for functions.  @Simn can confirm, though.

**Comment by Simn** [comment]
It works like this: In places where semicolons are usually required (after field declarations, after block-level expressions) they can be omitted if the last token was `}`.

This is why you can have:

```
var x = { field: "value" }
var y = x;
```

**Comment by as3boyan** [comment]
So we can add optional semicolon to class body `class A {};` but it doesn't work for expressions like `trace({x;};)`
